### PR TITLE
fixed download https://poe.com and create bot and edit bot and RuntimeError("Timed out waiting for other messages to send.")

### PR DESCRIPTION
**Pull request by canxin121**
#173 
all done
1.headers changed when creating a poe.client
2.key changed in creating bot and editing bot
3.fix RuntimeError("Timed out waiting for other messages to send.")
4.now you can either use old_handle or botid to edit the bot
5.some information in readme

### CODE REVIEW

**Comment by canxin121** [comment]
the same error? or not?

**Comment by lixiaohu1990** [comment]
Connecting to poe...
Failed to connect to poe due to failed to do request: Get "https://poe.com": net/http: request canceled (Client.Timeout exceeded while awaiting headers)

**Comment by lixiaohu1990** [comment]
Connecting to poe...
Failed to connect to poe due to failed to do request: Get "https://poe.com": net/http: request canceled (Client.Timeout exceeded while awaiting headers)

@canxin121

**Comment by lixiaohu1990** [comment]
also get error

 * Serving Flask app 'api'
 * Debug mode: off
INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:5100
 * Running on http://192.168.255.10:5100
INFO:werkzeug:Press CTRL+C to quit
Connecting to poe...
Failed to connect to poe due to failed to do request: Get "https://poe.com": EOF (Client.Timeout exceeded while awaiting headers)
INFO:werkzeug:127.0.0.1 - - [19/Jul/2023 15:35:13] "POST /add_token HTTP/1.1" 200 -

**Comment by canxin121** [comment]
image
try change the headers to your own browser's headers

**Comment by canxin121** [comment]
> that's great. it's worked !

what's your headers? give me a picture or the code

**Comment by lixiaohu1990** [comment]
> that's great. it's worked !

how to fix？ i stll error with

 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:5100
 * Running on http://192.168.255.10:5100
INFO:werkzeug:Press CTRL+C to quit
Connecting to poe...
Failed to connect to poe due to Timed out waiting for websocket to connect.
INFO:werkzeug:127.0.0.1 - - [19/Jul/2023 15:47:28] "POST /add_token HTTP/1.1" 200 -

**Comment by canxin121** [comment]
> > that's great. it's worked !
> 
> how to fix？ i stll error with
> 
> * Running on all addresses (0.0.0.0)
> * Running on http://127.0.0.1:5100
> * Running on http://192.168.255.10:5100
>   INFO:werkzeug:Press CTRL+C to quit
>   Connecting to poe...
>   Failed to connect to poe due to Timed out waiting for websocket to connect.
>   INFO:werkzeug:127.0.0.1 - - [19/Jul/2023 15:47:28] "POST /add_token HTTP/1.1" 200 -

https://github.com/ading2210/poe-api/pull/174#issuecomment-1641581141

**Comment by canxin121** [comment]
> no not same, my client_identifier is chrome111 Screenshot from 2023-07-19 15-06-05

Only Accept-Language is different from my latest code
mine is 'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6',

**Comment by lixiaohu1990** [comment]
> > > that's great. it's worked !
> > 
> > 
> > how to fix？ i stll error with
> > 
> > * Running on all addresses (0.0.0.0)
> > * Running on http://127.0.0.1:5100
> > * Running on http://192.168.255.10:5100
> >   INFO:werkzeug:Press CTRL+C to quit
> >   Connecting to poe...
> >   Failed to connect to poe due to Timed out waiting for websocket to connect.
> >   INFO:werkzeug:127.0.0.1 - - [19/Jul/2023 15:47:28] "POST /add_token HTTP/1.1" 200 -
> 
> [#174 (comment)](https://github.com/ading2210/poe-api/pull/174#issuecomment-1641581141)


i am replace the UA,but it is not work for me

image

image


image


image

**Comment by CaoYunzhou** [comment]
everyone give me headers text , i want copy it,pls thx

**Comment by canxin121** [comment]
the fix tested on my computer all runs smoothly, #try 

> everyone give me headers text , i want copy it,pls thx

can not use it with the headers of my pr ?

**Comment by CaoYunzhou** [comment]
> 

```
poe.headers = {
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:102.0) Gecko/20100101 Firefox/102.0",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,/;q=0.8",
    "Accept-Encoding": "gzip, deflate, br",
    "Accept-Language": "en-US,en;q=0.5",
    'Cache-Control': 'max-age=0',
    'Sec-Ch-Ua': '"Microsoft Edge";v="117", "Not;A=Brand";v="8", "Chromium";v="117"',
    'Sec-Ch-Ua-Mobile': '?0',
    'Sec-Fetch-Dest': 'document',
    'Sec-Fetch-Mode': 'navigate',
    'Sec-Fetch-Site': 'same-origin',
    'Sec-Fetch-User': '?1',
    "Upgrade-Insecure-Requests": "1"
        }
```

is not work , IP RG US-northamerica-northeast1-a

**Comment by gaoming0923** [comment]
ERROR:websocket:SSL not available. - goodbye
WARNING:root:Websocket closed with status None: None
ERROR:websocket:SSL not available. - goodbye
WARNING:root:Websocket closed with status None: None
ERROR:openai:Exception on /api/question [GET]
Traceback (most recent call last):
  File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/python/run/openai.py", line 14, in question
    client = poe.Client("ieJ9mtt75FB4oSHSLS_0SA%3D%3D")
  File "/home/python/run/poe.py", line 149, in __init__
    self.connect_ws()
  File "/home/python/run/poe.py", line 437, in connect_ws
    raise RuntimeError("Timed out waiting for websocket to connect.")
RuntimeError: Timed out waiting for websocket to connect.

**Comment by canxin121** [comment]
> ERROR:websocket:SSL not available. - goodbye WARNING:root:Websocket closed with status None: None ERROR:websocket:SSL not available. - goodbye WARNING:root:Websocket closed with status None: None ERROR:openai:Exception on /api/question [GET] Traceback (most recent call last): File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 2190, in wsgi_app response = self.full_dispatch_request() File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1486, in full_dispatch_request rv = self.handle_user_exception(e) File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1484, in full_dispatch_request rv = self.dispatch_request() File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1469, in dispatch_request return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args) File "/home/python/run/openai.py", line 14, in question client = poe.Client("ieJ9mtt75FB4oSHSLS_0SA%3D%3D") File "/home/python/run/poe.py", line 149, in **init** self.connect_ws() File "/home/python/run/poe.py", line 437, in connect_ws raise RuntimeError("Timed out waiting for websocket to connect.") RuntimeError: Timed out waiting for websocket to connect.

try change your headers to your own headers, and make sure that your proxy works

**Comment by gaoming0923** [comment]
> > 错误：websocket：SSL 不可用。- 再见警告：root：Websocket 已关闭，状态为无：无错误：websocket：SSL 不可用。- 再见警告：根：Websocket 已关闭，状态为无：无错误：openai：/api/question [GET] 上出现异常回溯（最近一次调用）：文件“/usr/local/python/lib/python3.10/site -packages/flask/app.py”，第 2190 行，在 wsgi_app 响应 = self.full_dispatch_request() 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第 1486 行，在 full_dispatch_request rv = self.handle_user_exception(e) 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第 1484 行，在 full_dispatch_request rv = self.dispatch_request() 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第1469行，在dispatch_request中返回self.ensure_sync(self.**init** self.connect_ws() 文件“/home/python/run/poe.py”，第 437 行，在 connect_ws 中引发 RuntimeError(“等待 websocket 连接超时。”) RuntimeError: 等待 websocket 连接超时。
> 
> 尝试将标头更改为您自己的标头，并确保您的代理可以正常工作

我的标头是自己的，

这是我的源码，麻烦您看下

**Comment by canxin121** [comment]
> > > 错误：websocket：SSL 不可用。- 再见警告：root：Websocket 已关闭，状态为无：无错误：websocket：SSL 不可用。- 再见警告：根：Websocket 已关闭，状态为无：无错误：openai：/api/question [GET] 上出现异常回溯（最近一次调用）：文件“/usr/local/python/lib/python3.10/site -packages/flask/app.py”，第 2190 行，在 wsgi_app 响应 = self.full_dispatch_request() 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第 1486 行，在 full_dispatch_request rv = self.handle_user_exception(e) 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第 1484 行，在 full_dispatch_request rv = self.dispatch_request() 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第1469行，在dispatch_request中返回self.ensure_sync(self.**init** self.connect_ws() 文件“/home/python/run/poe.py”，第 437 行，在 connect_ws 中引发 RuntimeError(“等待 websocket 连接超时。”) RuntimeError: 等待 websocket 连接超时。
> > 
> > 
> > 尝试将标头更改为您自己的标头，并确保您的代理可以正常工作
> 
> 我的标头是自己的，  这是我的源码，麻烦您看下

看下你的headers,我有点迷惑了,这poe到底在验证什么,还有确保代理能正常代理poe 的ws

**Comment by gaoming0923** [comment]
> > > > 错误：websocket：SSL 不可用。- 再见警告：root：Websocket 已关闭，状态为无：无错误：websocket：SSL 不可用。- 再见警告：根：Websocket 已关闭，状态为无：无错误：openai：/api/question [GET] 上出现异常回溯（最近一次调用）：文件“/usr/local/python/lib/python3.10/site -packages/flask/app.py”，第 2190 行，在 wsgi_app 响应 = self.full_dispatch_request() 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第 1486 行，在 full_dispatch_request rv = self.handle_user_exception(e) 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第 1484 行，在 full_dispatch_request rv = self.dispatch_request() 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第1469行，在dispatch_request中返回self.ensure_sync(self.**init** self.connect_ws() 文件“/home/python/run/poe.py”，第 437 行，在 connect_ws 中引发 RuntimeError(“等待 websocket 连接超时。”) RuntimeError: 等待 websocket 连接超时。
> > > 
> > > 
> > > 尝试将标头更改为您自己的标头，并确保您的代理可以正常工作
> > 
> > 
> > 我的标头是自己的，  这是我的源码，麻烦您看下
> 
> 看下你的headers,我有点迷惑了,这poe到底在验证什么,还有确保代理能正常代理poe 的ws

真尴尬，我刚接触Python，没看懂。。能说的通俗一点吗？

**Comment by canxin121** [comment]
> > > > > 错误：websocket：SSL 不可用。- 再见警告：root：Websocket 已关闭，状态为无：无错误：websocket：SSL 不可用。- 再见警告：根：Websocket 已关闭，状态为无：无错误：openai：/api/question [GET] 上出现异常回溯（最近一次调用）：文件“/usr/local/python/lib/python3.10/site -packages/flask/app.py”，第 2190 行，在 wsgi_app 响应 = self.full_dispatch_request() 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第 1486 行，在 full_dispatch_request rv = self.handle_user_exception(e) 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第 1484 行，在 full_dispatch_request rv = self.dispatch_request() 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第1469行，在dispatch_request中返回self.ensure_sync(self.**init** self.connect_ws() 文件“/home/python/run/poe.py”，第 437 行，在 connect_ws 中引发 RuntimeError(“等待 websocket 连接超时。”) RuntimeError: 等待 websocket 连接超时。
> > > > 
> > > > 
> > > > 尝试将标头更改为您自己的标头，并确保您的代理可以正常工作
> > > 
> > > 
> > > 我的标头是自己的，  这是我的源码，麻烦您看下
> > 
> > 
> > 看下你的headers,我有点迷惑了,这poe到底在验证什么,还有确保代理能正常代理poe 的ws
> 
> 真尴尬，我刚接触Python，没看懂。。能说的通俗一点吗？

你使用的poe库是我的fork的仓库里的吗? 是的话,你把我的git下载到本地修改了poe.py里面的headers了吗?                                 
然后代理问题,你本机用浏览器能直接访问poe.com吗

**Comment by gaoming0923** [comment]
> 是你仓库里的，poe.py里的headers改了，是我本地浏览器的。我本地浏览器可以访问poe.com。我本地运行时没问题的，报的错是我服务器上报的错，服务器是在阿里买的美国服务器

**Comment by canxin121** [comment]
> > 是你仓库里的，poe.py里的headers改了，是我本地浏览器的。我本地浏览器可以访问poe.com。我本地运行时没问题的，报的错是我服务器上报的错，服务器是在阿里买的美国服务器

可以在服务器登陆后抓取headers然后放进去试试

**Comment by gaoming0923** [comment]
> 

目前就是一个单个的服务，我通过postman调用的。。。

**Comment by canxin121** [comment]
可能跟header关系不大,主要我现在这里headers改成你们的也能使用,基本没有几个字母的差别,不知道为啥你们本机能用,服务器不可用,太怪了

**Comment by gaoming0923** [comment]
> 可能跟header关系不大,主要我现在这里headers改成你们的也能使用,基本没有几个字母的差别,不知道为啥你们本机能用,服务器不可用,太怪了

一直是websocket 
ERROR:websocket:SSL not available. - goodbye
WARNING:root:Websocket closed with status None: None

**Comment by gaoming0923** [comment]
> ERROR:websocket:SSL not available. - goodbye WARNING:root:Websocket closed with status None: None ERROR:websocket:SSL not available. - goodbye WARNING:root:Websocket closed with status None: None ERROR:openai:Exception on /api/question [GET] Traceback (most recent call last): File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 2190, in wsgi_app response = self.full_dispatch_request() File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1486, in full_dispatch_request rv = self.handle_user_exception(e) File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1484, in full_dispatch_request rv = self.dispatch_request() File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1469, in dispatch_request return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args) File "/home/python/run/openai.py", line 14, in question client = poe.Client("ieJ9mtt75FB4oSHSLS_0SA%3D%3D") File "/home/python/run/poe.py", line 149, in **init** self.connect_ws() File "/home/python/run/poe.py", line 437, in connect_ws raise RuntimeError("Timed out waiting for websocket to connect.") RuntimeError: Timed out waiting for websocket to connect.

这个问题，把poe里的websocket链接的wss改成ws就好了

**Comment by canxin121** [comment]
> > ERROR:websocket:SSL not available. - goodbye WARNING:root:Websocket closed with status None: None ERROR:websocket:SSL not available. - goodbye WARNING:root:Websocket closed with status None: None ERROR:openai:Exception on /api/question [GET] Traceback (most recent call last): File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 2190, in wsgi_app response = self.full_dispatch_request() File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1486, in full_dispatch_request rv = self.handle_user_exception(e) File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1484, in full_dispatch_request rv = self.dispatch_request() File "/usr/local/python/lib/python3.10/site-packages/flask/app.py", line 1469, in dispatch_request return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args) File "/home/python/run/openai.py", line 14, in question client = poe.Client("ieJ9mtt75FB4oSHSLS_0SA%3D%3D") File "/home/python/run/poe.py", line 149, in **init** self.connect_ws() File "/home/python/run/poe.py", line 437, in connect_ws raise RuntimeError("Timed out waiting for websocket to connect.") RuntimeError: Timed out waiting for websocket to connect.
> 
> 这个问题，把poe里的websocket链接的wss改成ws就好了

image
zh
指的是这个吗?zhzh

**Comment by noseon** [comment]
[canxin121](https://github.com/canxin121)your version correctly on google colab

**Comment by CaoYunzhou** [comment]
today everyboby how to resloved it bugs for WARNING:root:Server returned a status code of 403 while downloading https://poe.com. Retrying (5/10)...

**Comment by gaoming0923** [comment]
> > > 错误：websocket：SSL 不可用。- 再见警告：root：Websocket 已关闭，状态为无：无错误：websocket：SSL 不可用。- 再见警告：根：Websocket 已关闭，状态为无：无错误：openai：/api/question [GET] 上出现异常回溯（最近一次调用）：文件“/usr/local/python/lib/python3.10/site -packages/flask/app.py”，第 2190 行，在 wsgi_app 响应 = self.full_dispatch_request() 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第 1486 行，在 full_dispatch_request rv = self.handle_user_exception(e) 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第 1484 行，在 full_dispatch_request rv = self.dispatch_request() 文件“/usr/local/python/lib/python3.10/site-packages/flask/app.py”，第1469行，在dispatch_request中返回self.ensure_sync(self.**init** self.connect_ws() 文件“/home/python/run/poe.py”，第 437 行，在 connect_ws 中引发 RuntimeError(“等待 websocket 连接超时。”) RuntimeError: 等待 websocket 连接超时。
> > 
> > 
> > 这个问题，把poe里的websocket链接的wss改成ws就好了
> 
> 图像 zh 指的是这个吗？zhzh对的

**Comment by lixiaohu1990** [comment]
@canxin121 大佬，今天又出问题了。。。

**Comment by buaabarty** [comment]
I've updated to your latest code, and it still "Failed to download https://poe.com too many times". But your code was alright about ten hours ago.😢  @canxin121

**Comment by canxin121** [comment]
wait .i am checking on it

**Comment by CaoYunzhou** [comment]
截图_选择区域_20230720175647
this is worng，maybe nees proxy for poe.com

**Comment by CaoYunzhou** [comment]
截图_选择区域_20230720180128
i loss

File: README.md [hunk 1]
```diff
@@ -327,23 +327,23 @@ To use your browser's own headers, visit [this site](https://headers.uniqueostri
...
 ```python
 {
-  "Referrer": "https://poe.com/",
-  "Origin": "https://poe.com",
-  "Host": "poe.com",
-  "Sec-Fetch-Dest": "empty",
-  "Sec-Fetch-Mode": "cors",
-  "Sec-Fetch-Site": "same-origin"
+  'Cache-Control': 'max-age=0',
+  'Sec-Ch-Ua': '"Microsoft Edge";v="117", "Not;A=Brand";v="8", "Chromium";v="117"',
```
[the remark this answers is not in the document]
Comment (noseon) [reply README.md:341]: I see what happened was a generic blocking of the User-Agent, the right thing is to always use a new version of a browser

**Pull request merged**